### ISSUE TITLE
Missing Types for (undocumented) setComponentTemplate API

### ISSUE THREAD

**Issue by NullVoxPopuli** [typed-ember/ember-cli-typescript#1141]
**Edit by @chriskrycho 2022/02/10:** original title was "Missing Types for new APIs". Updated to reflect current state.

---

### Which package(s) does this problem pertain to?
  - [ ] @types/ember
  - [ ] @types/ember__string
  - [ ] @types/ember__polyfills
  - [x] @types/ember__object
  - [ ] @types/ember__utils
  - [ ] @types/ember__array
  - [ ] @types/ember__engine
  - [ ] @types/ember__debug
  - [ ] @types/ember__runloop
  - [ ] @types/ember__error
  - [ ] @types/ember__controller
  - [x] @types/ember__component
  - [ ] @types/ember__routing
  - [ ] @types/ember__application
  - [ ] @types/ember__test
  - [ ] @types/ember__test-helpers
  - [ ] @types/ember__service
  - [ ] @types/ember-data
  - [ ] @types/rsvp
  - [x] Other
  - [ ] I don't know

### What are instructions we can follow to reproduce the issue?

Try to use any of:

 - `import { dependentKeyCompat } from '@ember/object/compat';`
    - [docs](https://api.emberjs.com/ember/3.17/functions/@ember%2Fobject%2Fcompat/dependentKeyCompat)
 - `import { setComponentTemplate } from '@ember/component';`
    - [rfc](https://github.com/emberjs/rfcs/blob/master/text/0481-component-templates-co-location.md#low-level-primitives)
    - [code](https://github.com/emberjs/ember.js/blob/7763d7aa72445c88a0e63da800266465bef14ce1/packages/%40ember/-internals/glimmer/lib/utils/component-template.ts#L10)
 - all of ember-cli-htmlbars (but only `hbs` used)
    - [see issue](https://github.com/ember-cli/ember-cli-htmlbars/issues/468)

##### Reproduction Case
If you can, please try to fork [this codesandbox](https://codesandbox.io/s/github/mike-north/ember-new-output/tree/ts), and give us an example that demonstrates the problem. Paste the link below so that we can see what's going on

- Clone: https://github.com/NullVoxPopuli/emberclear/
- Checkout: `resolve-sidebar-performance-issues`
- cd packages/frontend/
- Delete:
  - types/@ember/
  - types/ember-cli-htmlbars
- in types/index.d.ts
  - Delete:
     ```ts
    declare module '@ember/component' {
      // TODO:  remove when this is actually a thing that exists?
      export function setComponentTemplate(template: string, klass: any): any;
    }
    ```
- Run `yarn lint:types`
  - notice type errors raised about 
    - `hbs` from `ember-cli-htmlbars`
    - `setComponentTemplate` from `@ember/component`
    - `dependentKeyCompat` from `@ember/object/compat`

```shell
tests/integration/routing/feature-flags-test.ts:4:21 - error TS7016: Could not find a declaration file for module 'ember-cli-htmlbars'. 'C:/Users/Preston/Documents/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/ember-cli-htmlbars/lib/index.js' implicitly has an 'any' type.
  Try `npm install @types/ember-cli-htmlbars` if it exists or add a new declaration (.d.ts) file containing `declare module 'ember-cli-htmlbars';`

4 import { hbs } from 'ember-cli-htmlbars';
                      ~~~~~~~~~~~~~~~~~~~~
```

```shell
app/components/pod/index/begin-button.ts:2:10 - error TS2614: Module '"../../../../../../../../../../../../Users/Preston/Documents/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/@types/ember__component"' has no exported member 'setComponentTemplate'. Did you mean to use 'import setComponentTemplate from "../../../../../../../../../../../../Users/Preston/Documents/Development/NullVoxPopuli/emberclear/packages/frontend/node_modules/@types/ember__component"' instead?

2 import { setComponentTemplate } from '@ember/component';
```

```shell
app/components/app/sidebar/chats/index.ts:4:36 - error TS2307: Cannot find module '@ember/object/compat'.

4 import { dependentKeyCompat } from '@ember/object/compat';
     
```

**Comment by NullVoxPopuli** [comment]
I am not really a fan of waiting to add types _until_ there are docs.

`setComponentTemplate` doesn't exist in ember (it's in `@glimmer/manager`), and going through all the loop holes to get docs to show up in the api docs for packages that aren't acutally in ember-source... _is a lot_. :cry: 

tldr: I don't know what to do.
adding:
```
// eslint-disable-next-line no-ts-whatever
// @ts-expect-error
```
everywhere for `setComponentTemplate`, `getValue`, and all the other new apis from 3.25+ is.. :weary:

**Comment by ef4** [comment]
FWIW, `setComponentTemplate` is a thoroughly RFC'd, very intentionally public API.

The docs have both technical and thematic limitations that make it challenging to categorize an API like `setComponentTemplate` which is public-but-low-level. I would suggest that written RFCs are equivalently good evidence to support policy choices about what to include in the types.

**Comment by chriskrycho** [comment]
Philosophically, I mostly agree!

Practically, there are very good reasons we haven’t allowed RFCs to be the source of truth for what we publish in our types, including that they historically haven’t necessarily matched what’s actually shipped! That’s become less a problem recently, as folks have tried to be more diligent about updating the RFC in that case, but even so, RFCs have been a poor source of truth for us.

(Here, I think with the review I did earlier today, we should be unblocked to publish the docs.)

**Comment by chriskrycho** [comment]
I believe this is now resolved since Ember ships its own types. :)
